File: src/main/java/com/dh/ondot/member/application/AuthFacade.java
package com.dh.ondot.member.application;

import com.dh.ondot.member.api.response.LoginResponse;
import com.dh.ondot.member.application.dto.Token;
import com.dh.ondot.member.domain.*;
import com.dh.ondot.member.domain.dto.UserInfo;
import com.dh.ondot.member.domain.enums.OauthProvider;
import com.dh.ondot.member.domain.service.MemberService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Slf4j
@Service
@RequiredArgsConstructor
public class AuthFacade {
    private final TokenFacade tokenFacade;
    private final OauthApiFactory oauthApiFactory;
    private final MemberService memberService;

    @Transactional
    public LoginResponse loginWithOAuth(OauthProvider oauthProvider, String accessToken) {
        OauthApi oauthApi = oauthApiFactory.getOauthApi(oauthProvider);
        UserInfo userInfo = oauthApi.fetchUser(accessToken);

        Member member = memberService.findOrRegisterOauthMember(userInfo, oauthProvider);

        boolean isNewMember = member.isNewMember();
        Token token = tokenFacade.issue(member.getId());
        if (isNewMember) {
            return LoginResponse.of(member.getId(),token.accessToken(), "", true);
        } else {
            return LoginResponse.of(member.getId(), token.accessToken(), token.refreshToken(), false);
        }
    }
}

File: src/main/kotlin/com/dh/ondot/member/application/AuthFacade.kt
package com.dh.ondot.member.application

import com.dh.ondot.member.presentation.response.LoginResponse
import com.dh.ondot.member.application.dto.Token
import com.dh.ondot.member.domain.OauthApiFactory
import com.dh.ondot.member.domain.enums.OauthProvider
import com.dh.ondot.member.domain.service.MemberService
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AuthFacade(
    private val tokenFacade: TokenFacade,
    private val oauthApiFactory: OauthApiFactory,
    private val memberService: MemberService,
) {
    private val log = LoggerFactory.getLogger(AuthFacade::class.java)

    @Transactional
    fun loginWithOAuth(oauthProvider: OauthProvider, accessToken: String): LoginResponse {
        val oauthApi = oauthApiFactory.getOauthApi(oauthProvider)
        val userInfo = oauthApi.fetchUser(accessToken)

        val member = memberService.findOrRegisterOauthMember(userInfo, oauthProvider)

        val isNewMember = member.isNewMember()
        val token: Token = tokenFacade.issue(member.id)
        return if (isNewMember) {
            LoginResponse.of(member.id, token.accessToken, "", true)
        } else {
            LoginResponse.of(member.id, token.accessToken, token.refreshToken, false)
        }
    }
}
